src/resources/nfts/index.ts

```diff
@@ -0,0 +1,170 @@
+      finished = true;
+    }
+
+    const currentNFTs =
+      queryClient.getQueryData<UniqueAsset[]>(nftsQueryKey({ address })) ?? [];
+
+    // iteratively update query data with new NFTs until the limit is hit
+    if (currentNFTs.length < NFTS_LIMIT) {
+      queryClient.setQueryData<UniqueAsset[]>(
+        nftsQueryKey({ address }),
+        cachedNFTs => uniqBy([...newNFTs, ...(cachedNFTs ?? [])], 'uniqueId')
+      );
+    }
+  }
+
+  // once we successfully fetch all NFTs, replace all cached NFTs with fresh ones
+  queryClient.setQueryData<UniqueAsset[]>(
+    nftsQueryKey({ address }),
+    () => freshNFTs
+  );
+
+  return freshNFTs;
+}
+
+export function useNFTs(): NFT[] {
+  // normal react query where we get new NFT formatted data
+  return [];
+}
+
+export function useLegacyNFTs(
+  address: string
+): { nfts: UniqueAsset[]; isLoading: boolean } {
+  const queryClient = useQueryClient();
+  const mounted = useIsMounted();
+
+  const [cursor, setCursor] = useState<string>();
+  const [isFinished, finish] = useReducer(() => true, false);
+  const [freshNFTs, setFreshNFTs] = useState<UniqueAsset[]>([]);
+
+  const queryKey = nftsQueryKey({ address });
+
+  // listen for query udpates
+  const query = useQuery<UniqueAsset[]>(queryKey, async () => [], {
+    enabled: false,
+    staleTime: Infinity,
+  });
+
+  const nfts = query.data ?? [];
+
+  useEffect(() => {
+    // stream in NFTs one simplehash response page at a time
+    const fetchNFTs = async () => {
+      const [simplehashResponse, polygonAllowlist] = await Promise.all([
+        fetchSimplehashNFTs(address, cursor),
+        fetchPolygonAllowlist(),
+      ]);
+
+      const { nfts: simplehashNFTs, nextCursor } = simplehashResponse;
+
+      const newNFTs = simplehashNFTs
+        .filter(nft => {
+          if (nft.chain === SimplehashChain.Polygon) {
+            return polygonAllowlist.includes(nft.contract_address);
+          }
+          return true;
+        })
+        .map(simplehashNFTToUniqueAsset);
+
+      const updatedFreshNFTs = [...newNFTs, ...freshNFTs];
+      setFreshNFTs(updatedFreshNFTs);
+
+      if (nextCursor && updatedFreshNFTs.length < NFTS_LIMIT) {
+        setCursor(nextCursor);
+      } else {
+        finish();
+      }
+
+      // iteratively update query data with new NFTs until the limit is hit
+      if (nfts.length < NFTS_LIMIT) {
+        queryClient.setQueryData<UniqueAsset[]>(queryKey, cachedNFTs =>
+          uniqBy([...newNFTs, ...(cachedNFTs ?? [])], 'uniqueId')
+        );
+      }
+    };
+    if (address && !isFinished && mounted.current) {
+      fetchNFTs();
+    }
+  }, [
+    address,
+    cursor,
+    freshNFTs,
+    isFinished,
+    mounted,
+    nfts.length,
+    queryClient,
+    queryKey,
+  ]);
+
+  useEffect(() => {
+    // once we successfully fetch all NFTs, replace all cached NFTs with fresh ones
+    if (isFinished) {
+      queryClient.setQueryData<UniqueAsset[]>(queryKey, () => freshNFTs);
+    }
```

Comment:
If we even need to use `uniqBy` we could either handle it incrementally above, or here.